libgalois/include/katana/MemoryManager.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+
+#include "katana/Manager.h"
+#include "katana/Result.h"
+#include "katana/config.h"
+
+namespace katana {
+// The memory manager singleton (MM).  Not thread safe.
+// The MM controls policy and does bookkeeping.  All memory allocation
+// is done by the system, mostly the C++ standard library.
+// The MM does not manage per-allocation tokens, it only manages sizes
+// Clients must call the proper functions or the MM will make
+// bad decisions.
+
+class KATANA_EXPORT MemoryManager {
+public:
+  MemoryManager(const MemoryManager&) = delete;
+  MemoryManager(MemoryManager&&) = delete;
+  MemoryManager& operator=(const MemoryManager&) = delete;
+  MemoryManager& operator=(MemoryManager&&) = delete;
+
+  // https://codereview.stackexchange.com/questions/173929/modern-c-singleton-template
+  static auto& MM() {
+    static MemoryManager mm_;
+    return mm_;
+  }
+
+  // Let the MM know about this manager.
+  void Register(Manager* manager);
+
+  // This manager is defunct.
+  // \p manager must have zero active and standby memory.
+  void Unregister(Manager* manager);
+
+  // Request \p bytes from the MM for active memory
+  // Application cannot continue if it does not get memory
```

Comment:
Then shouldn't the return value be a bool if it was successful or not? Am I missing something? 

Looking at its use, it's assumed that the allocation was successful before calling this. If so, please document that. Request makes it seem like it is requested before allocation.

Reply:
All of these functions are bookkeeping, not allocation.

So yes, allocation already succeeded, as will `BorrowActive`, so no bool return.

Reply:
I changed the documentation in this file to make this more clear.